Add Flux OCI documentation and example workflows (#1073)
fluxcd/website: content/en/docs/cheatsheets/oci-artifacts.md

{{% /alert %}}

### Consuming artifacts

On the server-side, Flux pulls OCI artifacts from container registries, extracts the Kubernetes manifests
and reconciles them on the cluster.

With [OCIRepository](/docs/components/source/ocirepositories/) you tell Flux which artifacts to pull
from a container registry and how to authenticate in order to download and make available
the artifacts' content inside the cluster.

The [OCIRepository](/docs/components/source/ocirepositories/) source can be used
in the same way as a [GitRepository](/docs/components/source/gitrepositories/).
You can be notified when new artifacts are pulled (Flux `Alert`),
you can trigger a pull with webhooks (Flux `Receiver`),
and you can apply the OCI artifact content on the cluster (Flux `Kustomization`).

### Helm OCI

For Helm users, Flux comes with [support](/docs/guides/helmreleases/#helm-oci-repository) for
defining Helm releases with charts stored in container registries.

Unlike Helm, the Flux OCI artifacts don't have a custom media type and they can be stored in
any container registry. The artifacts created by Flux can contain any type of configuration
besides Kubernetes manifests.

> The two artifacts types can be used together, for example you could bundle a `Namespace`,

Reviewer: ```suggestion
The two artifact types can be used together, for example you could bundle a `Namespace`,
```
I also wonder if it'd be better to say "source artifact", more explicit.

a `HelmRepository` and a `HelmRelease` into a Flux OCI artifact where the `HelmRepository`
points to a container registry where the chart is.

## Workflow examples

> What follows are examples for deploy a demo application to Kubernetes using manifests stored in Git.

Reviewer: maybe
```suggestion
Following are examples for deploying a demo application to Kubernetes using manifests stored in Git.
```

```mermaid
flowchart LR

> A((me)) --> B(Git Repository)

Reviewer: maybe
```suggestion
A((User)) --> B(Git Repository)
```
or even Developer?

B --> C((CI Job))
C --> D[Container Registry]
D --> E((Flux))
E --> F[Staging]
E --> G[Production]
```

We'll use the main branch to deploy to staging and Git tags for deploying on production.
When pushing changes to the main branch, a CI job generates the Kubernetes manifests for staging and
pushes them to a container registry using the FLux CLI. Flux running on the staging cluster
detects the new artifact digest, pulls the manifests and applies them.

> When cutting a release from them main branch by Git tagging a commit with a semver version, the CI job

Reviewer 2: ```suggestion
When cutting a release from the main branch by tagging a commit with a semver version, the CI job
```

```yaml
apiVersion: notification.toolkit.fluxcd.io/v1beta1
kind: Alert
metadata:
  name: podinfo
  namespace: flux-system
spec:
  summary: "podinfo in production"
  providerRef:
    name: slack
  eventSeverity: info
  eventSources:
    - kind: OCIRepository
      name: 'podinfo'
    - kind: Kustomization
      name: 'podinfo'
```

Or watch the Kubernetes events:

```console
$ kubectl alpha events -n flux-system --for OCIRepository/podinfo --watch

stored artifact with digest '04db795c5e8f039ee06e7f388e90ef9d16b713506dc100faed1773e0f8410d07'
from 'oci://ghcr.io/stefanprodan/manifests/podinfo',
origin source 'https://github.com/stefanprodan/podinfo.git',
origin revision '6.1.5/6b869d1a184969f7e24e6e4ad30be0b18f8b7416'
```

## Tracing

When publishing artifacts with `flux push` it is important to 
specify the Git source and revision with: 

```shell
flux push artifact oci://<repo url> --path=<manifests dir> \
	--source="$(git config --get remote.origin.url)" \
	--revision="$(git branch --show-current)/$(git rev-parse HEAD)"
```

The Git source and the revision are stored in the container registry as annotations in the OCI artifact manifest,

> and they are reflected in-cluster under the `status.artifact.metadata`.

Reviewer: maybe
```suggestion
and they are reflected in-cluster under the `status.artifact.metadata` of the `OCIRepository` object.
```
just to make it more clear.

change to Git.

### Helm OCI automation

Define an image repository and a semver policy for the Helm chart:

```yaml
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImageRepository
metadata:
  name: podinfo-chart
  namespace: flux-system
spec:
  image: ghcr.io/stefanprodan/charts/podinfo
  interval: 5m
---
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImagePolicy
metadata:
  name: podinfo-chart
  namespace: flux-system
spec:
  imageRepositoryRef:
    name: podinfo-chart
  policy:
    semver:
      range: 6.1.x
```

Then add the policy marker to the `HelmRelease` manifest in Git:

```yaml
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: podinfo
  namespace: flux-system
spec:
  interval: 10m
  targetNamespace: default
  chart:
    spec:
      chart: podinfo
      version: 6.1.0 # {"$imagepolicy": "flux-system:podinfo-chart:tag"}
      sourceRef:
        kind: HelmRepository
        name: podinfo
```

Based on the above configuration, Flux will scan the container registry every five minutes,
and when it finds a newer Helm chart version, it will update the

> `HelmRelease.spec.chart.spec.chart.version`and will push the change to Git.

Reviewer 3: ```suggestion
`HelmRelease.spec.chart.spec.chart.version` and will push the change to Git.
```
